### PR TITLE
Only allow to create replication2 database in MaintainerMode

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

*As we do not have capacity at the moment to do the necessary testing and stability phase, Replication2 is only available in MaintainerMode for a while and not publicly released.*

Effects of this PR:
1) When ArangoDB is compiled in NonMaintainer Mode:
  Starting ArangoDB Clusters without `--database.default-replication-version` will result in Replication Version 1.
  Starting ArangoDB Clusters with `--database.default-replication-version 1` will result in Replication Version 1.
  Starting ArangoDB Clusters with `--database.default-replication-version *` where * can be anything also 2, will result in
```
Error while processing command-line options for arangod:
  error setting value for option '--database.default-replication-version': invalid value '2'. Possible values: "1"
```

2) When ArangoDB is compiled in Maintainer Mode:
  Starting ArangoDB Clusters without `--database.default-replication-version` will result in Replication Version 1.
  Starting ArangoDB Clusters with `--database.default-replication-version 1` will result in Replication Version 1.
  Starting ArangoDB Clusters with `--database.default-replication-version 2` will result in Replication Version 2.
  Starting ArangoDB Clusters with `--database.default-replication-version *` where * can be anything except 1 or 2, will result in

```
Error while processing command-line options for arangod:
  error setting value for option '--database.default-replication-version': invalid value '3'. Possible values: "1", "2"
```

When trying to create a new Database in a Running Cluster the optional "replicationVersion" value is parsed and can be "1" or "2":

```
localhost:8530@_system> db._createDatabase("rep2", {replicationVersion: "3"})
JavaScript exception in file '/js/client/modules/@arangodb/arango-database.js' at 1104,5: ArangoError 10: Error parsing replicationVersion: Replication version must be "1" or "2", but is 3
```

in case it is set to "2" in non-maintainer version you will get an error:

```
db._createDatabase("hund", {replicationVersion: "2"})
JavaScript exception in file '/home/mchacki/arangodb/devel2/js/client/modules/@arangodb/arango-database.js' at 1104,5: ArangoError 9: Replication version 2 is disabled in this binary, cannot create replication version 2 databases.
!    throw new ArangoError(requestResult);
```

3) If you have existing data, containing Replication2 databases, and try to start a server in non-maintainer mode, both Coordinators and DBServers will die.
DBServers will not start as soon as they see Replication2 database persisted.
Coordinators with FATAL_EXIT as soon as they see a Replication2 database in the Plan.

4) If you restart the cluster changing the value of "--database.default-replication-version" in between restarts, the servers will refuse to start, as this would destroy the _system database.

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Cluster/ClusterInfo.cpp [hunk 1]
```diff
@@ -1073,6 +1074,24 @@ void ClusterInfo::loadPlan() {
...
+              << "Replication version 2 is disabled in this binary, but "
+                 "loading a "
+                 "version 2 database "
+              << "(named '" << info.getName() << "'). "
+              << "Creating such databases is disabled. Please dump the data, "
+                 "and "
+                 "recreate the database with replication version 1 (the "
+                 "default), "
+                 "and then restore the data.";
+          FATAL_ERROR_EXIT();
```
Comment (goedderz) [inline arangod/Cluster/ClusterInfo.cpp:1091]: How can we dump the data if the server isn't running? :)

File: arangod/Cluster/ClusterInfo.cpp [hunk 2]
```diff
@@ -1737,18 +1756,56 @@ void ClusterInfo::loadPlan() {
...
+                           "'--database.default-replication-version' between "
+                           "startups. (Was "
+                        << value.stringView() << ", is now set to "
+                        << replication::versionToString(
+                               systemDB->replicationVersion())
+                        << "). This would break the _system database. If you "
+                           "want your _system database to use a different "
+                           "replication version you need to start with an "
+                           "empty cluster and restore data.";
+                    FATAL_ERROR_EXIT();
```
Comment (goedderz) [inline arangod/Cluster/ClusterInfo.cpp:1806]: That doesn't sound right to me: I think the default replication version must never affect existing databases, only ever newly created ones. If it does, I think it's essential that we fix that (regardless of this PR).

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 3]
```diff
@@ -3010,6 +3010,11 @@ void RocksDBEngine::addSystemDatabase() {
   builder.add(StaticStrings::DatabaseName,
               VPackValue(StaticStrings::SystemDatabase));
   builder.add("deleted", VPackValue(false));
+  // Also store the ReplicationVersion when creating the Database
+  auto& df = server().getFeature<DatabaseFeature>();
+  builder.add(
+      StaticStrings::ReplicationVersion,
+      VPackValue(replication::versionToString(df.defaultReplicationVersion())));
```
Comment (goedderz) [inline arangod/RocksDBEngine/RocksDBEngine.cpp:3017]: This is good.

I'm wondering, what are we doing on load, if the replication version isn't set? We must always default to replication version 1, and not to defaultReplicationVersion there - your comment on the system database breaking above suggest that's not the case. Can we double-check this, please?

File: arangod/Cluster/ClusterInfo.cpp [hunk 4]
```diff
@@ -1737,18 +1756,56 @@ void ClusterInfo::loadPlan() {
           }
         }
 
-        // The systemDB does initially set the sharding attribute. Therefore,
-        // we need to set it here.
+        // The systemDB does initially set default values.
+        // As they can change with startup parameters we need
+        // to overwrite hem here to align with existing properties
```
Comment (apetenchea) [inline arangod/Cluster/ClusterInfo.cpp:1761]:
```suggestion
        // to overwrite them here to align with existing properties
```

File: arangod/RocksDBEngine/RocksDBEngine.cpp (hunk 3 above)
Comment (apetenchea) [reply arangod/RocksDBEngine/RocksDBEngine.cpp:3017]: As far as I know, that is defaulted to version 1 (I checked it in [VocbaseInfo.h](https://github.com/arangodb/arangodb/blob/devel/arangod/VocBase/VocbaseInfo.h#L178)). But I'll also let Michael answer this one.

**Review by apetenchea — COMMENTED** [review]
I am also slightly confused about the `ClusterInfo.cpp` part (same as Tobias' remark).

**Review by apetenchea — APPROVED** [review]
I don't think the changes required are massive. I'm also good with merging this in order to unblock Vitaly, and addressing the changes when Michael gets back.

I have created an issue here in case we decide to merge it: https://arangodb.atlassian.net/browse/CINFRA-886

**Pull request merged**